ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/navigation/NavigateToFilePresenter.java

```diff
@@ -93,22 +97,39 @@ public void onFileNameChanged(String fileName) {
             .createDto(ProjectSearchRequestDto.class)
             .withPath("")
             .withName(URL.encodePathSegment(fileName + "*"));
+    if (timer != null) {
+      timer.cancel();
+    }
 
-    requestTransmitter
-        .newRequest()
-        .endpointId(WS_AGENT_JSON_RPC_ENDPOINT_ID)
-        .methodName("project/search")
-        .paramsAsDto(requestParams)
-        .sendAndReceiveResultAsDto(ProjectSearchResponseDto.class, 20_000)
-        .onSuccess(response -> prepareResults(response))
-        .onFailure(error -> Log.error(getClass(), error.getMessage()))
-        .onTimeout(() -> Log.error(getClass(), "Project search request failed due timeout"));
+    timer =
+        new Timer() {
+          @Override
+          public void run() {
+            requestTransmitter
+                .newRequest()
+                .endpointId(WS_AGENT_JSON_RPC_ENDPOINT_ID)
+                .methodName("project/search")
+                .paramsAsDto(requestParams)
+                .sendAndReceiveResultAsDto(ProjectSearchResponseDto.class, 20_000)
+                .onSuccess(
+                    response -> {
+                      // Check that the file name from request corresponds to the actual file name
+                      // from the view.
+                      if (fileName.equals(view.getFileName())) {
+                        prepareResults(response);
+                      }
+                    })
+                .onFailure(error -> Log.error(getClass(), error.getMessage()))
+                .onTimeout(
+                    () -> Log.error(getClass(), "Project search request failed due timeout"));
+          }
+        };
+    timer.schedule(TYPING_PERIOD_DELAY_MS);
   }
 
   private void prepareResults(ProjectSearchResponseDto response) {
     List<SearchResultDto> results = response.getItemReferences();
-    sort(
-        results,
+    results.sort(
```

Comment:
Maybe it would be better to create static comparator?

Reply:
done